app/views/users/new.html.erb

```diff
@@ -1,3 +1,24 @@
 <% provide :title, t(".title") %>
 <h1><%= t ".h1" %></h1>
-<p><%= t ".p1" %></p>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user, url: signup_path do |f| %>
```

Comment:
nói cho c về những gì em tìm hiểu dược về form đi

Reply:
"The form_for helper is used to generate forms for interacting with Active Record objects."
Trái tim của trang signup là form để điền về đệ trình thông tin signup có liên quan trong các ô. Chúng ta có thể hoàn thành việc này trong Rails bằng cách sử dụng phương thức trợ giúp form_for, nó sẽ lấy đối tượng trong một Active Record và xây dựng 1 form sử dụng thuộc tính của đối tượng đó.
Đầu tiên ta tạo đối tượng User để làm đối số cho form_for (new action trong user_controller). Tiếp theo (như trong tut), ta áp dụng SCSS để thay đổi kiểu dáng cho form.
Để hiểu rõ form được định nghĩa trong đoạn code, ta sẽ chia nó thành các phần nhỏ. Trước hết ta xét cấu trúc bên ngoài của nó, chứa cặp thẻ đóng mở embedded Ruby giúp gọi và thực thi form_for bên trong. Ở phía trong, từ khóa "do" biểu thị rằng form_for nhận 1 block chứa 1 biến "f"(viết tắt của "form"). Như thường lệ với trường hợp của Rails helpers, chúng ta không cần biết quá nhiều về việc thực thi, nhưng ta cái ta cần biết là đối tượng f làm gì: khi gọi 1 hàm tương ứng với thành phần HTML form - như là trường text, nút radio, hoặc trường mật khẩu - f trả về code cho element đó được thiết kế đặc biệt để đặt một thuộc tính của đối tượng @user. Nói cách khác, 
<%= f.label :name %>
<%= f.text_field :name %>
tạo ra những gì HTML cần để làm nên một element trường text có gắn nhãn phù hợp với việc thiết đặt tên thuộc tính của 1 User model. Đoạn code nhúng trên sẽ được chuyển thành dạng HTML như sau:
<label for="user_name">Name</label>
<input id="user_name" name="user[name]" type="text" />
Tùy theo nhãn trong code nhúng (và khi được dịch sang HTML) form sẽ được đối xử khác nhau (ví dụ với nhãn :email và type="email" trên một số thiết bị di động sẽ hiện ra bàn phím đặc biệt được tối ưu hóa cho việc nhập email).

Reply:
cấu trúc đầy đủ của form gồm có những gì?

Reply:
form_for là một phương thức trong helper để tạo ra đoạn mã html nhằm mục đích xây dựng thẻ <form></form> (Listing 7.15 dịch qua HTML tại Listing 7.17 chap7).
tham số đầu tiên truyền vào form_for là một record, là một biến instance sinh ra từ trong controller @user nó sẽ thông qua biến đó để set action cho form_for. Phần url thay vì truyền path theo quy ước, có thể set controller và action thủ công bằng cách sử dụng {action: :index, controller: :statics} chẳng hạn, truyền cụ thể action nào thuộc controller nào. 
biến f đưa ra một khối đối thuộc FormBuiler hiện thực hóa đối tượng model @user, đối tượng mà được truyền vào form_for. từ biến f có thể sử dụng các phương thức FormHelper để tạo ra các thẻ input, checkbox,...
tạo thẻ input text: <%= f.text_field :name %>
tạo thẻ input password: <%= f.password_field :password %>
tạo thẻ checkbox: <%= f.check_box :check_box %>
thẻ cực kỳ quan trong đó là submit: <%= f.submit "Create my account" %>
...còn rất nhiều thuộc tính nữa.
nếu muốn thêm các id, class cho các thẻ input chỉ cần thêm thuộc tính: <%= f.text_field :name, class: 'form-control' %>

Reply:
` thông qua biến đó để set action cho form_for.` cụ thể là gì?

Và câu trả lời của của em vẫn  chưa đúng trọng tâm nhé

Reply:
ta có action new trong controller
def new
   @user = User.new
 end
form_for tương ứng khi với view new.html.erb
<%= form_for @user, url: signup_path do |f| %>
<%= render "shared/error_messages", user: @user %>

  <%= f.label :name %>
  <%= f.text_field :name, class: "form-control" %>

  <%= f.label :email %>
  <%= f.email_field :email, class: "form-control" %>

  <%= f.label :password %>
  <%= f.password_field :password, class: "form-control" %>

  <%= f.label :password_confirmation, t(".password_confirmation") %>
  <%= f.password_field :password_confirmation, class: "form-control" %>

  <%= f.submit t(".submit"), class: "btn btn-primary" %>
<% end %>

@user chính là object thực được đang được chỉnh sửa. Thiết đặt định tuyến truyền vào :url hash. form_for method gọi 1 trình dựng form cho đối tượng (biến f). Method để tạo trình điều khiển form được gọi trên trình dựng form cho đối tượng f. HTML đầu ra là:

<form accept-charset="UTF-8" action="/users" class="new_user"
      id="new_user" method="post">
  <input name="utf8" type="hidden" value="&#x2713;" />
  <input name="authenticity_token" type="hidden"
         value="NNb6+J/j46LcrgYUC60wQ2titMuJQ5lLqyAbnbAUkdo=" />
  <label for="user_name">Name</label>
  <input id="user_name" name="user[name]" type="text" />

  <label for="user_email">Email</label>
  <input id="user_email" name="user[email]" type="email" />

  <label for="user_password">Password</label>
  <input id="user_password" name="user[password]"
         type="password" />

  <label for="user_password_confirmation">Confirmation</label>
  <input id="user_password_confirmation"
         name="user[password_confirmation]" type="password" />

  <input class="btn btn-primary" name="commit" type="submit"
         value="Create my account" />
</form>
Sau khi điền form và submit thành công sẽ thực hiện action create và show, không thành công sẽ render lại new

Reply:
- 3 loại form: form_with, form_for, form_tag
- Option full: form_for(record, options = {}, &block)
	+ as: Thay đổi tiền tố các trường
	+ url:  nơi mà form được gửi đến
	+ method: tên phương thức
	+ html: chỉnh sửa trực tiếp các thuộc tính của form
- Sự khác nhau giữa các loại form:
	+ form_tag: Tạo form nhưng không sử dụng đối tượng nào trong database
		+ eg: <%= form_tag users_path do %>
  			<%= text_field_tag :email %>
  			<%= submit_tag %>
		      <% end %>
	+ form_for: Tạo form sử dụng đối tượng (có thể là 1 record đã tồn tại hoặc 1 đối tượng mới) trong database 
		+ eg: <%= form_for @user do |form| %>
  			<%= form.text_field :email %>
  			<%= form.submit %>
		      <% end %>
	+ form_with: dùng cho cả khi không dùng đối tượng hoặc có đối tượng, không tự động thêm id và class như form_for và form_tag
Thêm 1 ddiefu nữa là form_for cung cấp 1 method để check việc tạo mới hay update